cty/function/stdlib/collection.go

```diff
@@ -299,20 +299,52 @@ var ContainsFunc = function.New(&function.Spec{
 		},
 	},
 	Type: function.StaticReturnType(cty.Bool),
-	Impl: func(args []cty.Value, retType cty.Type) (ret cty.Value, err error) {
+	Impl: func(args []cty.Value, retType cty.Type) (cty.Value, error) {
 		arg := args[0]
 		ty := arg.Type()
 
 		if !ty.IsListType() && !ty.IsTupleType() && !ty.IsSetType() {
 			return cty.NilVal, errors.New("argument must be list, tuple, or set")
 		}
 
-		_, err = Index(cty.TupleVal(arg.AsValueSlice()), args[1])
-		if err != nil {
-			return cty.False, nil
+		if args[0].IsNull() {
+			return cty.NilVal, errors.New("cannot search a nil list or set")
+		}
+
+		if args[0].LengthInt() == 0 {
+			return cty.NilVal, errors.New("cannot search an empty list or set")
+		}
+
+		if !args[0].IsKnown() || !args[1].IsKnown() {
+			return cty.UnknownVal(cty.Bool), nil
+		}
+
+		containsUnknown := false
+		for it := args[0].ElementIterator(); it.Next(); {
+			_, v := it.Element()
+			// if we don't find a match, but have an unknown element, we need
+			// to return unknown
+			if !v.IsKnown() {
```

Comment:
This special case doesn't really hurt, and I'm okay with keeping it, but just wanted to note that `eq := args[1].Equals(v)` followed by `eq.IsKnown()` would've been sufficient to cach this, because `Equals` is an "operation method" rather than an "integration method", as defined [in the _`cty` concepts_ doc](https://github.com/zclconf/go-cty/blob/master/docs/concepts.md#the-cty-go-api).

Reply:
Thanks! I will clean this up since I want to make the change above too.